lib/python/picongpu/picmi/diagnostics/binning.py

```diff
@@ -0,0 +1,191 @@
+            # We first add mass and momentum
+            # and later use their symbols inside of the same preamble.
+            self.get("mass")
+            self.get("momentum")
+            if attribute == "gamma":
+                symbols = sympy.Symbol("gamma")
+            if attribute == "kinetic energy":
+                symbols = sympy.Symbol("Ekin")
+            if attribute == "velocity":
+                symbols = sympy.symbols("vx,vy,vz")
+            self.used_attributes |= {symbols: attribute}
+
+        else:
+            symbols = sympy.Symbol(attribute)
+            self.used_attributes |= {symbols: attribute}
+
+        return symbols
+
+
+@typeguard.typechecked
+class BinningFunctor:
+    def check(self):
+        pass
+
+    def __init__(
+        self,
+        name: str,
+        functor: Callable[[BinningParticle], Any],
+        return_type: type | str,
+    ):
+        self.name = name
+        self.functor = functor
+        self.return_type = return_type
+
+    def get_as_pypicongpu(self) -> PyPIConGPUBinningFunctor:
+        self.check()
+        particle = BinningParticle()
+        functor_expression = self.functor(particle)
+        return PyPIConGPUBinningFunctor(
+            name=self.name,
+            functor_expression=functor_expression,
+            attribute_mapping=particle.get_attribute_map(),
+            return_type=self.return_type,
+        )
+
+
+@typeguard.typechecked
+class BinSpec:
+    def __init__(self, kind, start, stop, nsteps):
+        self.kind = kind
+        self.start = start
+        self.stop = stop
+        self.nsteps = nsteps
+
+    def get_as_pypicongpu(self):
+        return PyPIConGPUBinSpec(self.kind.lower().capitalize(), self.start, self.stop, self.nsteps)
+
+
+@typeguard.typechecked
+class BinningAxis:
+    def __init__(
+        self,
+        functor: BinningFunctor,
+        bin_spec: BinSpec,
+        name: str | None = None,
+        use_overflow_bins: bool = True,
+    ):
+        self.functor = functor
+        self.bin_spec = bin_spec
+        self.name = name or functor.name
+        self.use_overflow_bins = use_overflow_bins
+
+    def get_as_pypicongpu(self) -> PyPIConGPUBinningAxis:
+        return PyPIConGPUBinningAxis(
+            name=self.name,
+            functor=self.functor.get_as_pypicongpu(),
+            bin_spec=self.bin_spec.get_as_pypicongpu(),
+            use_overflow_bins=self.use_overflow_bins,
+        )
+
+
+@typeguard.typechecked
+class Binning:
+    def __init__(
+        self,
+        name: str,
+        deposition_functor: BinningFunctor,
+        axes: list[BinningAxis],
+        species: PICMISpecies | list[PICMISpecies],
+        period: Optional[TimeStepSpec] = None,
+        openPMD: Optional[dict] = None,
+        openPMDExt: Optional[str] = None,
+        openPMDInfix: Optional[str] = None,
+        timeAveraging: bool = True,
+        dumpPeriod: int = 1,
```

Comment:
probably a dumb question what happens in the binning plugin if dumpPeriod < period ?